# Exercise-Part1

# 1. Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

# 2. Import data

In [2]:
# Defining path
path=r'C:\Users\ponni\Achievement 4'

In [3]:
# Import customer data
df_customer = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [4]:
df_customer

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374  
...              ...        ...     ...  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095  

[206209 rows x 10 columns]

#### There are lot of column name errors. So the column name's should be renamed.

# 3. Data Wrangling

In [5]:
# Renaming the columns
df_customer.columns = ["user_id", "first_name", "last_name", "gender", "state", "age", "date_joined", "no_of_dependants", "marital_status", "income($)"]

In [6]:
df_customer

user_id first_name last_name  gender           state  age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        no_of_dependants marital_status  income($)  
0                      3        married     165665  
1                      0         single      59285  
2                      2        married      99568  
3                      0         single      42049  
4                      1        married      40374  
...                  ...            ...        ...  
206204                 1        married     148828  
206205                 3        married     168639  
206206                 2        married      53374  
206207                 1        married      99799  
206208                 1        married      57095  

[206209 rows x 10 columns]

In [7]:
# check the data type
df_customer.describe()

user_id            age  no_of_dependants      income($)
count  206209.000000  206209.000000     206209.000000  206209.000000
mean   103105.000000      49.501646          1.499823   94632.852548
std     59527.555167      18.480962          1.118433   42473.786988
min         1.000000      18.000000          0.000000   25903.000000
25%     51553.000000      33.000000          0.000000   59874.000000
50%    103105.000000      49.000000          1.000000   93547.000000
75%    154657.000000      66.000000          3.000000  124244.000000
max    206209.000000      81.000000          3.000000  593901.000000

#### The user_id column is needed for calculation. So, we can change the data type to string and the column wont appear in future while using the describe function.

In [8]:
# changing data type for user_id, as we dont need for calculation
df_customer['user_id'] = df_customer['user_id'].astype('str')

In [9]:
df_customer.dtypes

user_id             object
first_name          object
last_name           object
gender              object
state               object
age                  int64
date_joined         object
no_of_dependants     int64
marital_status      object
income($)            int64
dtype: object

# 4. Data consistecy checks

In [10]:
# 1. Finding missing values
df_customer.isnull().sum()

user_id                 0
first_name          11259
last_name               0
gender                  0
state                   0
age                     0
date_joined             0
no_of_dependants        0
marital_status          0
income($)               0
dtype: int64

In [11]:
# viewing missing data
df_nan = df_customer[df_customer['first_name'].isnull() == True]

In [12]:
df_nan

user_id first_name last_name  gender         state  age date_joined  \
53       76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73       13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82       89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99       96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105      29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...        ...        ...       ...     ...           ...  ...         ...   
206038  121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044  200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090  167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162  187532        NaN     Floyd  Female    California   39    4/1/2020   
206171  116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        no_of_dependants marital_status  income($)  
53                     2        married      41709  
73                     0         single      82518  
82                     3        married     117099  
99                     1        married     155673  
105                    3        married     151819  
...                  ...            ...        ...  
206038                 3        married      87783  
206044                 2        married     108488  
206090                 1        married      45275  
206162                 0         single      56325  
206171                 2        married      59222  

[11259 rows x 10 columns]

In [13]:
# checking for duplicates
df_customer[df_customer.duplicated()]

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, no_of_dependants, marital_status, income($)]
Index: []

#### There are no duplicates found and there are 11259 missing values in the first_name column. Hence, it can't be deleted as it might affect the analysis. We can leave the missing values as such now.

In [14]:
# 2. Addressing missing data types
for col in df_customer.columns.tolist():
  weird = (df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customer[weird]) > 0:
    print (col)

first_name


In [15]:
# changing the column first_name data type
df_customer['first_name'] = df_customer['first_name'].astype('str')

In [16]:
# checking the column dtype
df_customer['first_name'].dtype

dtype('O')

In [17]:
# checking again for mixed data type 
for col in df_customer.columns.tolist():
  weird = (df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customer[weird]) > 0:
    print (col)

#### The mixed data type problem has been solved.

# 5. Combing data

In [18]:
# importing prepared data
ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared data', 'orders_products_merged_4.pkl'))

In [19]:
ords_prods_merged.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                   999.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...    price_range_loc     busiest day        busiest_days  \
0         Soda  ...  Mid-range product  Regularly busy  Regulary busy days   
1         Soda  ...  Mid-range product  Regularly busy        Slowest days   
2         Soda  ...  Mid-range product  Regularly busy        Slowest days   
3         Soda  ...  Mid-range product      Least busy        Slowest days   
4         Soda  ...  Mid-range product      Least busy        Slowest days   

  busiest_period_of_day max_order  loyalty_flag average_price spending_flag  \
0           Most orders        10  New customer      6.367797   Low spender   
1        Average orders        10  New customer      6.367797   Low spender   
2           Most orders        10  New customer      6.367797   Low spender   
3        Average orders        10  New customer      6.367797   Low spender   
4           Most orders        10  New customer      6.367797   Low spender   

   order_frequency_flag customer_frequency_flag  
0                  21.0   Non frequent customer  
1                  21.0   Non frequent customer  
2                  21.0   Non frequent customer  
3                  21.0   Non frequent customer  
4                  21.0   Non frequent customer  

[5 rows x 24 columns]

In [20]:
# to see all the column names
print(ords_prods_merged.columns.tolist())

['order_id', 'user_id', 'order_number', 'order_dow', 'order_hour_of_day', 'days_since_prior_order', 'product_id', 'add_to_cart_order', 'reordered', 'product_name', 'aisle_id', 'department_id', 'prices', '_merge', 'price_range_loc', 'busiest day', 'busiest_days', 'busiest_period_of_day', 'max_order', 'loyalty_flag', 'average_price', 'spending_flag', 'order_frequency_flag', 'customer_frequency_flag']


In [21]:
ords_prods_merged.drop('_merge', axis=1)

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0          2539329        1             1          2                  8   
1          2398795        1             2          3                  7   
2           473747        1             3          3                 12   
3          2254736        1             4          4                  7   
4           431534        1             5          4                 15   
...            ...      ...           ...        ...                ...   
32404854   1320836   202557            17          2                 15   
32404855     31526   202557            18          5                 11   
32404856    758936   203436             1          2                  7   
32404857   2745165   203436             2          3                  5   
32404858   3093936   205420             1          4                 14   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                          999.0         196                  1          0   
1                           15.0         196                  1          1   
2                           21.0         196                  1          1   
3                           29.0         196                  1          1   
4                           28.0         196                  1          1   
...                          ...         ...                ...        ...   
32404854                     1.0       43553                  2          1   
32404855                     3.0       43553                  2          1   
32404856                   999.0       42338                  4          0   
32404857                    15.0       42338                 16          1   
32404858                   999.0       28818                  8          0   

                           product_name  ...    price_range_loc  \
0                                  Soda  ...  Mid-range product   
1                                  Soda  ...  Mid-range product   
2                                  Soda  ...  Mid-range product   
3                                  Soda  ...  Mid-range product   
4                                  Soda  ...  Mid-range product   
...                                 ...  ...                ...   
32404854            Orange Energy Shots  ...  Low-range product   
32404855            Orange Energy Shots  ...  Low-range product   
32404856          Zucchini Chips, Pesto  ...  Mid-range product   
32404857          Zucchini Chips, Pesto  ...  Mid-range product   
32404858  Hot Oatmeal Multigrain Raisin  ...  Mid-range product   

             busiest day        busiest_days busiest_period_of_day max_order  \
0         Regularly busy  Regulary busy days           Most orders        10   
1         Regularly busy        Slowest days        Average orders        10   
2         Regularly busy        Slowest days           Most orders        10   
3             Least busy        Slowest days        Average orders        10   
4             Least busy        Slowest days           Most orders        10   
...                  ...                 ...                   ...       ...   
32404854  Regularly busy  Regulary busy days           Most orders        31   
32404855  Regularly busy  Regulary busy days           Most orders        31   
32404856  Regularly busy  Regulary busy days        Average orders         3   
32404857  Regularly busy        Slowest days         Fewest orders         3   
32404858      Least busy        Slowest days           Most orders        16   

              loyalty_flag average_price  spending_flag order_frequency_flag  \
0             New customer      6.367797    Low spender                 21.0   
1             New customer      6.367797    Low spender                 21.0   
2             New customer      6.367797    Low spender                 21.0   
3             New customer      6.367797    Low spender                 21.0   
4     

In [34]:
ords_prods_merged_2= ords_prods_merged.drop(['order_frequency_flag','customer_frequency_flag','add_to_cart_order', 'reordered'], axis=1)

MemoryError: Unable to allocate 2.17 GiB for an array with shape (9, 32404859) and data type int64

In [23]:
ords_prods_merged_2.shape

(32404859, 20)

In [28]:
# For combining the data the common column is user_id
# checking for the data type in ords_prods_merged_2
ords_prods_merged_2['user_id'].dtype

dtype('int64')

In [29]:
# changing the user_id to string to not affect the calculation
ords_prods_merged_2['user_id'] = ords_prods_merged_2['user_id'].astype('str')
ords_prods_merged_2['user_id'].dtype

dtype('O')

In [31]:
# merging the two dataframes
df_merged = ords_prods_merged_2.merge(df_customer, on = 'user_id')

MemoryError: Unable to allocate 1.93 GiB for an array with shape (8, 32404859) and data type int64

# 6. Exporting data

In [35]:
df_customer.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'df_customer_final.pkl'))

In [36]:
ords_prods_merged_2.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged_5.pkl'))

MemoryError: 

# Continuing to merge data in fresh file